File: 算法/Python/第1章 算法概述/1.3 基础算法/monte_pi.py
import random
import math

def montePI(n):
    """
        蒙特卡洛算法计算PI
        Args:
            n (int): 样本数量
    """
    cnt = 0
    for _ in range(n):
        x = random.random()     # [0, 1)
        y = random.random()     # [0, 1)
        if math.sqrt(x**2 + y**2) <= 1:
            cnt += 1
    return 4.0 * cnt / n

def main():
    print(montePI(10000000))

if __name__ == "__main__":
    main()

File: 算法/C/第1章 算法概述/1.3 基础算法/monte_pi.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

/**
 * @brief  蒙特卡洛算法计算PI
 * @param  n: 样本数量
 * @retval PI的近似值
 */
double montePI(int n) {
    int cnt = 0;        // 圆内点的数量
    for(int i = 0; i < n; i++) {
        double x = rand() / (RAND_MAX + 1.0);   // [0, 1]
        double y = rand() / (RAND_MAX + 1.0);   // [0, 1]
        if(sqrt(x*x + y*y) <= 1) {
            cnt++;
        }
    }
    return 4.0 * cnt / n;
}

int main() {
    srand(time(NULL));
    printf("PI = %f\n", montePI(10000000));
    return 0;
}